### PR TITLE
feat(core): make the CfnParameter class mutable

### LINKED ISSUE #9364: [cfn-include] Make CfnParameters mutable
Currently, CfnParameters are not mutable once they are created.

### Proposed Solution

I will be opening a PR to address this.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by comcalvi**
Closes #9364 

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

### CODE REVIEW

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 1]
```diff
@@ -107,17 +117,166 @@ export class CfnParameter extends CfnElement {
...
+  public set allowedPattern(pattern: string | undefined) {
+    this._allowedPattern = pattern;
+  }
+
+  /**
+   * An array containing the list of values allowed for the parameter.
+   *
+   * @default - No constraints on values allowed for parameter.
+   */
+
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:182]: Unnecessary empty line.

File: packages/@aws-cdk/cloudformation-include/README.md [hunk 2]
```diff
@@ -146,6 +146,24 @@ and any changes you make to it will be reflected in the resulting template:
 output.value = cfnBucket.attrArn;
 ```
 
+## Parameters
```
Comment (skinny85) [inline packages/@aws-cdk/cloudformation-include/README.md:149]: Can you actually move this section to be right above the existing "Conditions" section?

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 3]
```diff
@@ -107,17 +117,166 @@ export class CfnParameter extends CfnElement {
...
+  /**
+   * A numeric value that determines the smallest numeric value you want to allow for Number types.
+   *
+   * @default - None.
+   */
+  public get minValue(): number | undefined {
+    return this._minValue;
+  }
+
+  public set minValue(len: number | undefined) {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:267]: JSDocs for all setters are missing - surprised our linter did not pick this up, but should be updated nonetheless!

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 4]
```diff
@@ -107,17 +117,166 @@ export class CfnParameter extends CfnElement {
...
 
   /**
    * Indicates if this parameter is configured with "NoEcho" enabled.
    */
   public get noEcho(): boolean {
-    return !!this.props.noEcho;
+    return !!this._noEcho;
+  }
+
+  public set noEcho(echo: boolean) {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:278]: I believe the setter should take `boolean | undefined` (if you just want to rely on the CFN default).

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 5]
```diff
@@ -141,8 +300,8 @@ export class CfnParameter extends CfnElement {
    * The parameter value, if it represents a string list.
    */
   public get valueAsList(): string[] {
-    if (!isListType(this.type)) {
-      throw new Error(`Parameter type (${this.type}) is not a string list type`);
+    if (!isListType(this._type)) {
+      throw new Error(`Parameter type (${this._type}) is not a string list type`);
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:304]: How about we just leave it as `this.type`?

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 6]
```diff
@@ -151,8 +310,8 @@ export class CfnParameter extends CfnElement {
    * The parameter value, if it represents a number.
    */
   public get valueAsNumber(): number {
-    if (!isNumberType(this.type)) {
-      throw new Error(`Parameter type (${this.type}) is not a number type`);
+    if (!isNumberType(this._type)) {
+      throw new Error(`Parameter type (${this._type}) is not a number type`);
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:314]: Same comment as above ("How about we just leave it as `this.type`?").

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 7]
```diff
@@ -164,17 +323,17 @@ export class CfnParameter extends CfnElement {
...
+          Default: this._default,
+          AllowedPattern: this._allowedPattern,
+          AllowedValues: this._allowedValues,
+          ConstraintDescription: this._constraintDescription,
+          Description: this._description,
+          MaxLength: this._maxLength,
+          MaxValue: this._maxValue,
+          MinLength: this._minLength,
+          MinValue: this._minValue,
+          NoEcho: this._noEcho,
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:336]: I would probably use the public getters here, I don't see why not.

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 8]
```diff
@@ -107,17 +117,165 @@ export class CfnParameter extends CfnElement {
...
+  public set minValue(len: number | undefined) {
+    this._minValue = len;
   }
 
   /**
    * Indicates if this parameter is configured with "NoEcho" enabled.
    */
-  public get noEcho(): boolean {
-    return !!this.props.noEcho;
+  public get noEcho(): boolean | undefined {
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:273]: Just to be clear, I meant only the setter. The getter should stay unchanged.

File: packages/@aws-cdk/core/lib/cfn-parameter.ts [hunk 9]
```diff
@@ -165,16 +323,16 @@ export class CfnParameter extends CfnElement {
...
+          Default: this.default,
+          AllowedPattern: this.allowedPattern,
+          AllowedValues: this.allowedValues,
+          ConstraintDescription: this.constraintDescription,
+          Description: this.description,
+          MaxLength: this.maxLength,
+          MaxValue: this.maxValue,
+          MinLength: this.minLength,
+          MinValue: this.minValue,
+          NoEcho: this.noEcho,
```
Comment (skinny85) [inline packages/@aws-cdk/core/lib/cfn-parameter.ts:335]: Here, do use `this._noEcho` (you need the triple state boolean here).

File: packages/@aws-cdk/core/lib/cfn-parameter.ts (hunk 8 above)
Comment (comcalvi) [reply packages/@aws-cdk/core/lib/cfn-parameter.ts:273]: Their types have to be the same though. We can't have one being `boolean | undefined` and the other being `boolean`

Comment (skinny85) [reply packages/@aws-cdk/core/lib/cfn-parameter.ts:273]: Hmmm, that sucks then.

OK, leave the setter as `boolean` then. If we need to, we can add a separate method like `unsetNoEcho()` later.

**Pull request merged**